### PR TITLE
[Clang] Remove `IDNS_Ordinary` flag in `IndirectFieldDecl::IdentifierNamespace`

### LINKED ISSUE #31295: Invalid cast to CXXRecordDecl in Sema::BuildPossibleImplicitMemberExpr
|  |  |
| --- | --- |
| Bugzilla Link | [31947](https://llvm.org/bz31947) |
| Version | trunk |
| OS | All |
| Attachments | [Testcase](https://user-images.githubusercontent.com/2278807/143755097-2a66b4f5-a832-40a8-bcfb-034fe3317951.gz) |

## Extended Description 

The command 

$ clang durbin-aac882.c

```c
nion w {
  {};
  struct {
    igned t}typedef union f[t
```

crashes with

clang-5.0: /root/src/llvm/include/llvm/Support/Casting.h:236: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = clang::CXXRecordDecl; Y = clang::DeclContext; typename llvm::cast_retty<X, Y*>::ret_type = clang::CXXRecordDecl*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
#&#8203;0 0x00000000016bd69a llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/root/build/llvm/release/bin/clang-5.0+0x16bd69a)
#&#8203;1 0x00000000016bb33e llvm::sys::RunSignalHandlers() (/root/build/llvm/release/bin/clang-5.0+0x16bb33e)
#&#8203;2 0x00000000016bb4b2 SignalHandler(int) (/root/build/llvm/release/bin/clang-5.0+0x16bb4b2)
#&#8203;3 0x00007f876bff1390 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x11390)
#&#8203;4 0x00007f876ad25428 gsignal /build/glibc-t3gR2i/glibc-2.23/signal/../sysdeps/unix/sysv/linux/raise.c:54:0
#&#8203;5 0x00007f876ad2702a abort /build/glibc-t3gR2i/glibc-2.23/stdlib/abort.c:91:0
#&#8203;6 0x00007f876ad1dbd7 __assert_fail_base /build/glibc-t3gR2i/glibc-2.23/assert/assert.c:92:0
#&#8203;7 0x00007f876ad1dc82 (/lib/x86_64-linux-gnu/libc.so.6+0x2dc82)
#&#8203;8 0x00000000027ec78d clang::Sema::BuildPossibleImplicitMemberExpr(clang::CXXScopeSpec const&, clang::SourceLocation, clang::LookupResult&, clang::TemplateArgumentListInfo const*, clang::Scope const*) (/root/build/llvm/release/bin/clang-5.0+0x27ec78d)
#&#8203;9 0x0000000002764c48 clang::Sema::ActOnIdExpression(clang::Scope*, clang::CXXScopeSpec&, clang::SourceLocation, clang::UnqualifiedId&, bool, bool, std::unique_ptr<clang::CorrectionCandidateCallback, std::default_delete<clang::CorrectionCandidateCallback> >, bool, clang::Token*) (/root/build/llvm/release/bin/clang-5.0+0x2764c48)
#&#8203;10 0x00000000023dca41 clang::Parser::ParseCastExpression(bool, bool, bool&, clang::Parser::TypeCastState) (/root/build/llvm/release/bin/clang-5.0+0x23dca41)
#&#8203;11 0x00000000023dea0d clang::Parser::ParseCastExpression(bool, bool, clang::Parser::TypeCastState) (/root/build/llvm/release/bin/clang-5.0+0x23dea0d)
#&#8203;12 0x00000000023deaa9 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) (/root/build/llvm/release/bin/clang-5.0+0x23deaa9)
#&#8203;13 0x00000000023bd19f clang::Parser::ParseBracketDeclarator(clang::Declarator&) (/root/build/llvm/release/bin/clang-5.0+0x23bd19f)
#&#8203;14 0x00000000023bd61b clang::Parser::ParseMisplacedBracketDeclarator(clang::Declarator&) (/root/build/llvm/release/bin/clang-5.0+0x23bd61b)
#&#8203;15 0x00000000023c5ba7 clang::Parser::ParseDirectDeclarator(clang::Declarator&) (/root/build/llvm/release/bin/clang-5.0+0x23c5ba7)
#&#8203;16 0x00000000023bc8e4 clang::Parser::ParseDeclaratorInternal(clang::Declarator&, void (clang::Parser::*)(clang::Declarator&)) (/root/build/llvm/release/bin/clang-5.0+0x23bc8e4)
#&#8203;17 0x00000000023bcdc2 clang::Parser::ParseDeclarator(clang::Declarator&) (/root/build/llvm/release/bin/clang-5.0+0x23bcdc2)
#&#8203;18 0x00000000023c0962 clang::Parser::ParseStructDeclaration(clang::ParsingDeclSpec&, llvm::function_ref<void (clang::ParsingFieldDeclarator&)>) (/root/build/llvm/release/bin/clang-5.0+0x23c0962)
#&#8203;19 0x00000000023c11f5 clang::Parser::ParseStructUnionBody(clang::SourceLocation, unsigned int, clang::Decl*) (/root/build/llvm/release/bin/clang-5.0+0x23c11f5)
#&#8203;20 0x00000000023d7d9d clang::Parser::ParseClassSpecifier(clang::tok::TokenKind, clang::SourceLocation, clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, bool, clang::Parser::DeclSpecContext, clang::Parser::ParsedAttributesWithRange&) (/root/build/llvm/release/bin/clang-5.0+0x23d7d9d)
#&#8203;21 0x00000000023b7c68 clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*) (/root/build/llvm/release/bin/clang-5.0+0x23b7c68)
#&#8203;22 0x000000000239d006 clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/root/build/llvm/release/bin/clang-5.0+0x239d006)
#&#8203;23 0x000000000239d7f1 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) [clone .part.160] [clone .constprop.185] (/root/build/llvm/release/bin/clang-5.0+0x239d7f1)
#&#8203;24 0x000000000239d84f clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/root/build/llvm/release/bin/clang-5.0+0x239d84f)
#&#8203;25 0x000000000239f62f clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (/root/build/llvm/release/bin/clang-5.0+0x239f62f)
#&#8203;26 0x00000000023a3382 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) (/root/build/llvm/release/bin/clang-5.0+0x23a3382)
#&#8203;27 0x00000000023a361c clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) (/root/build/llvm/release/bin/clang-5.0+0x23a361c)
#&#8203;28 0x000000000239881d clang::ParseAST(clang::Sema&, bool, bool) (/root/build/llvm/release/bin/clang-5.0+0x239881d)
#&#8203;29 0x0000000001f99d6a clang::CodeGenAction::ExecuteAction() (/root/build/llvm/release/bin/clang-5.0+0x1f99d6a)
#&#8203;30 0x0000000001bba266 clang::FrontendAction::Execute() (/root/build/llvm/release/bin/clang-5.0+0x1bba266)
#&#8203;31 0x0000000001b8cb76 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/root/build/llvm/release/bin/clang-5.0+0x1b8cb76)
#&#8203;32 0x0000000001c41e3a clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/root/build/llvm/release/bin/clang-5.0+0x1c41e3a)
#&#8203;33 0x00000000009efd88 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/root/build/llvm/release/bin/clang-5.0+0x9efd88)
#&#8203;34 0x00000000009a03dc main (/root/build/llvm/release/bin/clang-5.0+0x9a03dc)
#&#8203;35 0x00007f876ad10830 __libc_start_main /build/glibc-t3gR2i/glibc-2.23/csu/../csu/libc-start.c:325:0
#&#8203;36 0x00000000009ec5d9 _start (/root/build/llvm/release/bin/clang-5.0+0x9ec5d9)

The file durbin-aac882.c (reduced using afl-tmin) is attached. clang trunk r294933.

### PR DESCRIPTION
**Pull request by zwuis**
There is a `IDNS_Ordinary` flag in `IndirectFieldDecl::IdentifierNamespace` so that members in nested anonymous struct/union can be found as ordinary identifiers.

```c
struct S {
  struct { int x; };
  // Previous behaviour: `x` in previous line is found
  // Expected: nothing is found
  int arr[sizeof(x)];
};
```

This PR fixes this issue.

Fixes #31295.

### CODE REVIEW

**Review by Sirraide — COMMENTED** [review]
Looks reasonable, but I’d prefer it if someone more familiar w/ the C standard than me could take another look at this.
CC @AaronBallman

**Review by AaronBallman — APPROVED** [review]
The C standard does not make this particularly clear, but the changes are correct.

Per C23 6.2.1p4 Each member of the structure has file scope (because why not) but per 6.2.3p1, each member is in a distinct name space and the member is disambiguated by the type of the expression used to access the member via the `.` or `->` operator. That last bit is why the member cannot be an ordinary identifier within the structure itself.

So LGTM, thank you!

**Comment by zwuis** [comment]
Thank you for your review!

I don't have commit access. Please help me merge this PR if it's ready.

**Comment by cor3ntin** [comment]
@AaronBallman @Sirraide Do we have tests for C++? https://godbolt.org/z/bdvrdKso1

**Comment by Sirraide** [comment]
> @AaronBallman @Sirraide Do we have tests for C++? https://godbolt.org/z/bdvrdKso1

I don’t believe the standard supports anonymous structs in C++, but it seems we’re handling them the same way as GCC, i.e. the code in the godbolt link is accepted in C++ mode by both us and GCC.

That said, I don’t know whether we have tests for this off the top of my head, but I should hope so considering that we support the feature...

**Comment by Sirraide** [comment]
Well, looks like I spoke too soon, because I can’t find any C++ tests for this particular case

**Comment by zwuis** [comment]
> I don’t believe the standard supports anonymous structs in C++

Do we need tests about anonymous union? C++ standard supports it.

**Comment by AaronBallman** [comment]
> > I don’t believe the standard supports anonymous structs in C++
> 
> Do we need tests about anonymous union? C++ standard supports it.

Yes, we should add tests for C++ as well; good call @cor3ntin and @Sirraide!

**Pull request merged**
